web/src/settings_profile_fields.ts

```diff
@@ -430,15 +484,21 @@ function open_edit_form_modal(e) {
                 .addClass("display_in_profile_summary_tooltip disabled_label");
         }
 
-        if (Number.parseInt(field.type, 10) === field_types.SELECT.id) {
+        if (field.type === field_types.SELECT.id) {
             set_up_select_field_edit_form($profile_field_form, field);
         }
 
-        if (Number.parseInt(field.type, 10) === field_types.EXTERNAL_ACCOUNT.id) {
+        if (field.type === field_types.EXTERNAL_ACCOUNT.id) {
+            const external_account_data =
+                settings_components.external_account_field_schema.parse(field_data);
```

Comment:
Is it okay to parse data this frequently with `Zod` just to make TypeScript happy, even though we're confident it should never fail unless there's a bug in the code, are there any better approaches?
(I've done this on multiple instances)

Reply:
Yeah, I think it is totally fine. Infact it is one of `zod`' s attributes ( to bring untyped data into the typed world.)

Reply:
@afeefuddin As a general thumb rule, any data that is coming from outside world e.g server should be parsed with `zod`.
To quote Zulip's [documentation](https://zulip.readthedocs.io/en/latest/testing/mypy.html#when-is-a-type-annotation-justified) on this:
> keep in mind that a **type annotation should always represent a guarantee, not an aspiration**. If you have validated that some value is an int, it can go in an int annotated variable.

So always validate untyped data coming from outside with `zod`.

Reply:
Yes, but this data is already typed. Here I am parsing it again just to get rid of the union type. 
For example, if we have a type `foo | bar` and we are sure it should be of type `foo`, should we parse it using Zod?

Reply:
In that case you should ideally have a _discriminator_ key that helps typescript directly infer the type. For ex. both `foo` and `bar` has a field called `xyz` (_discriminator key_) . Now from the type `foo | bar` `TypeScript` would be able to infer the particular type ( ie. `foo` or `bar`) depending on value of `xyz`.

Reply:
Yep, I had the same thought initially that's why I was confirming about Zod parsing on whether is it a good approach to parse it like this or  should we refactor the types with an additional field.

Reply:
One approach which I am thinking of can be using an initial of `unknown` type during `JSON.parse` and then parsing data later on within `if` conditions since we don't really have a discriminatory key and data is completely coming from the server. The only thing which discriminates these two objects is the key `type` in the `field` object.

Reply:
I would have typed FieldData something like ...
```
type FieldData= Record<string,unknown>
```
instead of the union and later parsed it using `zod` as you did. But what you suggested is correct too.